hushline/premium.py

```diff
@@ -0,0 +1,536 @@
+
+
+def get_subscription(user: User) -> stripe.Subscription | None:
+    if user.stripe_subscription_id is None:
+        return None
+
+    return stripe.Subscription.retrieve(user.stripe_subscription_id)
+
+
+def handle_subscription_created(subscription: stripe.Subscription) -> None:
+    # customer.subscription.created
+
+    user = db.session.query(User).filter_by(stripe_customer_id=subscription.customer).first()
+    if user:
+        user.stripe_subscription_id = subscription.id
+        user.stripe_subscription_status = StripeSubscriptionStatusEnum(subscription.status)
+        user.stripe_subscription_cancel_at_period_end = subscription.cancel_at_period_end
+        user.stripe_subscription_current_period_end = datetime.fromtimestamp(
+            subscription.current_period_end
+        )
+        user.stripe_subscription_current_period_start = datetime.fromtimestamp(
+            subscription.current_period_start
+        )
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with customer ID {subscription.customer}")
+
+
+def handle_subscription_updated(subscription: stripe.Subscription) -> None:
+    # customer.subscription.updated
+
+    # If subscription changes to cancel or unpaid, downgrade user
+    user = db.session.query(User).filter_by(stripe_subscription_id=subscription.id).first()
+    if user:
+        user.stripe_subscription_status = StripeSubscriptionStatusEnum(subscription.status)
+        user.stripe_subscription_cancel_at_period_end = subscription.cancel_at_period_end
+        user.stripe_subscription_current_period_end = datetime.fromtimestamp(
+            subscription.current_period_end
+        )
+        user.stripe_subscription_current_period_start = datetime.fromtimestamp(
+            subscription.current_period_start
+        )
+
+        current_app.logger.info("status is: " + subscription.status)
+        if subscription.status in ["active", "trialing"]:
+            user.tier_id = BUSINESS_TIER
+        else:
+            user.tier_id = FREE_TIER
+
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_subscription_deleted(subscription: stripe.Subscription) -> None:
+    # customer.subscription.deleted
+
+    user = db.session.query(User).filter_by(stripe_subscription_id=subscription.id).first()
+    if user:
+        user.tier_id = FREE_TIER
+        user.stripe_subscription_id = None
+        user.stripe_subscription_status = None
+        user.stripe_subscription_cancel_at_period_end = None
+        user.stripe_subscription_current_period_end = None
+        user.stripe_subscription_current_period_start = None
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_invoice_created(invoice: stripe.Invoice) -> None:
+    # invoice.created
+
+    try:
```

Comment:
Catch the `ValueError` only around the constructor?

Reply:
Fixed in ff5f2173ca30559812e5b4a9652b6355b061200a